ion-schema/src/isl/version_based_isl/mod.rs

```diff
@@ -0,0 +1,480 @@
+    /// Represents an id for the given ISL model
+    id: String,
+    /// Represents the schema header if its present which contains imports and any open content provided in the header.
+    /// For ISL 2.0 schema header also contains user reserved fields.
+    /// If the schema header is not present this is None.
+    schema_header: Option<IslSchemaHeader<V>>,
+    /// Represents the schema footer if its present which contains any open content provided in the footer.
+    /// If the schema footer is not present this is None.
+    schema_footer: Option<IslSchemaFooter>,
+    /// Represents all the IslType defined in this schema file.
+    /// For more information: https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#type-definitions
+    types: Vec<IslType<V>>,
+    /// Represents all the inline IslImportTypes in this schema file.
+    inline_imported_types: Vec<IslImportType>,
+    /// Represents open content as `Element`s
+    /// Note: This doesn't preserve the information about where the open content is placed in the schema file.
+    /// e.g. This method doesn't differentiate between following schemas with open content:
+    /// ```ion
+    /// $foo
+    /// $bar
+    /// type::{ name: foo, codepoint_length: 1 }
+    /// ```
+    ///
+    /// ```ion
+    /// type::{ name: foo, codepoint_length: 1 }
+    /// $foo
+    /// $bar
+    /// ```
+    open_content: Vec<Element>,
+    _version: PhantomData<V>,
+}
+
+impl<V: IslVersionTrait> IslSchema<V> {
+    pub fn new<A: AsRef<str>>(
+        id: A,
+        types: Vec<IslType<V>>,
+        inline_imports: Vec<IslImportType>,
+        open_content: Vec<Element>,
+    ) -> Self {
+        Self {
+            id: id.as_ref().to_owned(),
+            schema_header: None,
+            schema_footer: None,
+            types,
+            inline_imported_types: inline_imports,
+            open_content,
+            _version: Default::default(),
+        }
+    }
+
+    pub fn with_schema_header(self, schema_header: IslSchemaHeader<V>) -> Self {
+        Self {
+            schema_header: Some(schema_header),
+            ..self
+        }
+    }
+
+    pub fn with_schema_footer(self, schema_footer: IslSchemaFooter) -> Self {
+        Self {
+            schema_footer: Some(schema_footer),
+            ..self
+        }
+    }
+
+    pub fn id(&self) -> String {
+        self.id.to_owned()
+    }
+
+    pub fn types(&self) -> &[IslType<V>] {
+        &self.types
+    }
+
+    pub fn inline_imported_types(&self) -> &[IslImportType] {
+        &self.inline_imported_types
+    }
+
+    /// Provides top level open content for given schema
+    /// For open content defined within type definitions use IslType#open_content()
+    pub fn open_content(&self) -> &Vec<Element> {
+        &self.open_content
+    }
+
+    /// Provides the schema header value if present,
+    /// Otherwise returns None.
+    pub fn schema_header(&self) -> &Option<IslSchemaHeader<V>> {
+        &self.schema_header
+    }
```

Comment:
If you return a `&Option<_>`, the user can't `unwrap()` it because the `Option` itself is borrowed. Use `Option::as_ref()` to get an `Option<&_>` instead.
```suggestion
    pub fn schema_header(&self) -> Option<&IslSchemaHeader<V>> {
        self.schema_header.as_ref()
    }
```